accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,493 @@
+...
+  List<int> x = ...;
+  var v = x.best();
+  List<num> y = ...;
+  var w = y.best();
+```
+
+Here all three extensions apply to both invocations.
+
+For `x.best()`, the most specific one is `BoxList`. Because `Box<List<int>>` is a proper subtype of both ` Box<iterable<int>>` and `Box<List<num>>`, we expect `BoxList` to be the best implementation. The return type causes `v` to have type `int`. If we had chosen `BoxSpec` instead, the return type could only be `num`, which is why we choose the most specific instantiated type as the winner. 
+
+For `y.best()`, the most specific extension is `BoxSpec`. The instantiated `on` types that are compared are `Box<Iterable<num>>` for `BoxCom` and `Box<List<num>>` for the two other. Using the instantiate-to-bounds types as tie-breaker, we find that `Box<List<Object>>` is less precise than `Box<List<num>>`, so the code of `BoxSpec` has more precise information available for its method implementation. The type of `w` becomes `Box<List<num>>`.
+
+In practice, unintended extension method name conflicts are likely to be rare. Intended conflicts happen where the same author is providing more specialized versions of an extension for subtypes, and in that case, picking the extension which has the most precise types available to it is considered the best choice.
+
+### Overriding Access
+
+If two or more extensions apply to the same member access, or if a member of the receiver type takes precedence over an extension method, or if the extension is imported with a prefix, then it is possible to force an extension member invocation:
+
+```dart
+MyList(object).quickSort();
+```
+
+or if you don't want the type argument to the extension to be inferred:
+
+```dart
+MyList<String>(object).quickSort();
+```
+
+or if you imported the extension with a prefix to avoid name collision:
+
+```dart
+prefix.MyList<String>(object).quickSort();
+```
+
+The syntax looks like a constructor invocation, but it does not create a new object.
+
+If `object.quickSort()` would invoke an extension method of `MyList`, then `MyList(object).quickSort()` will invoke the exact same method in the same way.
+
+This mode of invocation also allows invocation of extensions that are *not* in scope, perhaps as `prefix.MyExtension(object).extensionMember()` if imported with a prefix.
+
+The syntax is not *convenient*&mdash;you have to put the "constructor" invocation up front, which removes the one advantage that extension methods have over normal static methods. It is not intended as the common use-case, but as an escape hatch out of conflicts.
+
+### Static Members and Member Resolution
+
+Static member declarations in the extension declaration can be accessed the same way as static members of a class or mixin declaration: By prefixing with the extension's name.
+
+Example:
+
+```dart
+extension MySmart on Object {
+  smart() => smartHelper(this);  // valid
+  static smartHelper(Object o) { ... }
+}
+...
+  MySmart.smartHelper(someObject);  // valid
```

Comment:
Do we disallow references to generic type parameters in static methods?  Otherwise we need to allow `MyGenericSmart<int>.smartHelper(someObject)`.

Reply:
Yeah, Dart already forbids that. Static members exist on the *class's* namespace, and don't hang off the instantiated type or have access to the class's type arguments.